Two-factor authentication

jrauh01 opened the pull request:
This PR is based on #5397 and will therefore close it. Introduces basic two-factor authentication (**2FA**) using Time-based One-Time Passwords (**TOTP**) support for Icinga Web 2.

TOTP (Time-based One-Time Password) is a type of two-factor authentication (2FA) method. It generates short-lived numeric codes (usually 6-8 digits) based on:
1. A shared secret key (known to both the server and your authenticator app).
2. The current time (in 30-second intervals by default).

Since the code changes frequently and is only valid for a short window, it makes accounts harder to hack, even if someone steals your password.

After enabling 2FA, users authenticate with their usual Icinga Web credentials and are then prompted to provide a valid TOTP code from their authenticator app (e.g., Google Authenticator, Authy).

We will use the defaults of **6 digit tokens**, **30 seconds interval** and **sha-1 algorithm** for token generation. The generated token is valid for **10 seconds before and after the current time** to allow some clock drift.

### Current Authentication Process

Currently, Icinga Web authentication follows this workflow:
- First it checks if there is set a remember me cookie from past logins.
- If so, it attempts to log the user in using the encrypted credentials stored in the cookie.
- If there is no remember me cookie set, it will show a login form.
- If submitted it will compare the password with the hash in the database.
- If correct it will setup the user and store it in the database.
- Additionally it will create the remember me cookie if the slider for that in the login form is checked.
- Then it will redirect to the dashboard or another target set by url param `redirect`.

### 2FA Implementation

This implementation is split in two main parts.
1. The configuration of 2FA via the account settings
2. The actual authentication process with token input

#### Configuration

The configuration of TOTP-based 2FA is fully integrated into the user’s account settings. A new form (`TwoFactorConfigForm`) allows users to enable 2FA, verify, or remove a TOTP secret directly from their account settings.

When a user with the required permission (`user/two-factor-authentication`) accesses their account, the `AccountController` loads the current TOTP secret from the database (`TwoFactorTotp::loadFromDb()`). If none exists, a new secret is generated (`TwoFactorTotp::generate()`) and passed into the form.

The form (`TwoFactorConfigForm`) then guides the user through the following workflow:

- **Enable 2FA**: If no TOTP secret is stored, the user can opt in via a checkbox. Once enabled, a new secret is created, and the form displays both a QR code (rendered by `TwoFactorTotp::createQRCode()`) and a manual provisioning URI. The user can scan this code with an authenticator app or copy the URI via a **copy-to-clipboard** element.
- **Verification Step**: To prevent misconfiguration, the user must enter a valid TOTP code generated by their app. The code is checked using the verify method of the `TwoFactorTotp` class. Only if the verification succeeds will the secret be persisted to the database (`saveToDb()`).
- **Remove 2FA**: If a secret already exists in the database, the form instead provides the option to remove it. On form submit, the stored secret is deleted (`removeFromDb()`). **(Should be extended in the future to ask for confirmation!)**

#### Authentication

Once a user has a TOTP secret stored, the login flow gains a second step. After the username and password are verified, the session sets a `2fa_must_challenge_token` flag, stores a temporary user object, and if “Stay logged in” was checked, a temporary remember-me cookie in the session. The remember-me cookie is not persisted yet. This prevents bypassing 2FA by simply reusing the cookie. `Auth::isAuthenticated()` also enforces that a user is only considered fully logged in once the TOTP challenge has succeeded, unless explicitly skipped for trusted cookies.

The `LoginForm` is rewritten now extending `ipl\Web\Compat\CompatForm` and decides which elements it should render. If the `2fa_must_challenge_token` flag is set in the session, the elements to verify the TOTP token are displayed, otherwise the normal login form elements. A valid remember-me cookie can bypass both, the password login and the token challenge: if present and verified, the cookie is renewed, persisted, and the user is logged in.

When the `LoginForm` is submitted via the `btn_submit_verify_2fa` submit button, the secret is loaded from the database and the entered code is verified. A successful check sets `$twoFactorSuccessful` to `true` on the user object that was temporarily stored in the session after password login, clears the `2fa_must_challenge_token` session flag, persists any deferred remember-me cookie, and calls `AuthenticationHook::triggerLogin()`. Submitting the form via the `btn_submit_cancel_2fa` button lets users abort, which purges the session and resets the login.

### Future

Possible extensions for the future could be:
- Encrypt secrets in the database
- Option for administrators to enforce 2fa for users
- Ask for password verification to add or remove a TOTP secret in the account settings

Requires GD php extension for QR code generation
Requires https://github.com/Icinga/icinga-php-thirdparty/pull/53
Closes #5397

lippserd committed:
wip

Jan-Schuppik committed:
Initial implementation

From https://github.com/Icinga/icingaweb2/pull/5397

jrauh01 committed:
Remove `mtime` column from database

The column is never used and won't in the future.

jrauh01 committed:
Rename stuff

- 'TotpForm' to 'TotpConfigForm'
- 'Authentication\Totp' to 'IcingaTotp'
- 'Model\Totp' to 'TotpModel'

jrauh01 committed:
Move 'PsrClock' to Authentication stuff in library

jrauh01 committed:
Change title for 2fa settings

jrauh01 committed:
Add missing types in `User`

jrauh01 committed:
Refactor 2FA secret handling

Previously, a new 2FA secret was generated and stored temporarily in the
session. It was only persisted to the database when the user clicked the
"Save Changes" button. This behavior has been removed.

Now, the secret is written directly to the database once it has been
initially verified with a token. To generate a new secret if one
already exists, just remove the old one. There is no longer a separate
setting to toggle whether 2FA is enabled for a user — 2FA is considered
enabled if a secret exists for that user in the database.

jrauh01 committed:
Adjust comment

jrauh01 committed:
Adjust 'Cancel' button

Add cancel submit button explicitly instead of setting it implicitly
by calling `setSubmitLabel('...')` to add `.btn-cancel` class so the
cancel button does not look like a primary button.

jrauh01 committed:
Rename text input to 'token'

jrauh01 committed:
Rename session key for 2fa challenge

Rename from 'must_challenge_2fa_token' to '2fa_must_challenge_token'
for a uniform naming schema in the future always starting with '2fa_'.

jrauh01 committed:
Remove session cookie for successful 2fa

Remove the cookie for a successful 2fa, because there is no need for it.
Instead set the 2fa directly to successful on the user.

jrauh01 committed:
Verify 2fa token properly

jrauh01 committed:
Simplify if condition

jrauh01 committed:
Adjust remember me functionality to work with 2fa

If 2fa is enable the remember me cookie only gets set if the 2fa
authentication was successful. To log the user back in from the
cookie the 2fa will be skipped.

Al2Klimov commented:
> Requires https://github.com/Icinga/icinga-php-thirdparty/pull/53

I'm afraid I have to convert this PR to draft in this case.

> The failed phpcs tests are not due to my changes.

Well, the current master is all-✅.

Al2Klimov commented on library/Icinga/Model/TotpModel.php:
```diff
@@ -0,0 +1,38 @@
+
+class TotpModel extends Model
+{
+
```
```suggestion
```

Al2Klimov commented on library/Icinga/Model/TotpModel.php:
```diff
@@ -0,0 +1,38 @@
+
+    /**
+     * @inheritDoc
+     */
```
@lippserd @nilmerg Correct me if anything, but IIRC, we don't use such anymore.

Al2Klimov commented on application/forms/Account/TotpConfigForm.php:
```diff
@@ -0,0 +1,217 @@
+ *
+ * This form is used to manage the TOTP settings of a user account.
+ */
+class TotpConfigForm extends Form
```
An ipl form makes sense, unless this is planned to be used as a subform in established stuff.

jrauh01 replied:
Would you apply this only to the `TotpConfigForm`? Or to the `Cancel2FAForm` and `Challenge2FAForm` as well?

Al2Klimov commented on application/forms/Account/TotpConfigForm.php:
```diff
@@ -0,0 +1,217 @@
+                'text',
+                'user_totp_secret',
+                [
+                    'value' => $dbTotp->getSecret()
```
Why'd you show an existing secret to the user on re-display of the form?

IMAO showing it makes sense only once for usage as recovery code.

jrauh01 replied:
Yes, this was a temporary solution for development.

Al2Klimov commented on application/forms/Account/TotpConfigForm.php:
```diff
@@ -0,0 +1,217 @@
+            $this->addElement(
+                'text',
+                'user_totp_secret',
+                [
```
Why should I as a user generate and provide my own secret?

Al2Klimov commented on schema/mysql.schema.sql:
```diff
@@ -64,5 +64,12 @@ CREATE TABLE icingaweb_schema (
   CONSTRAINT idx_icingaweb_schema_version UNIQUE (version)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=DYNAMIC;
 
+CREATE TABLE `icingaweb_totp`(
```
Even if I imagine for a moment I had a schema upgrade, when I log in as admin to apply it via web UI, IW2 greets me with:

> SQLSTATE[42S02]: Base table or view not found: 1146 Table 'iw2.icingaweb_totp' doesn't exist

Al2Klimov commented on application/forms/Account/TotpConfigForm.php:
```diff
@@ -0,0 +1,217 @@
+ *
+ * This form is used to manage the TOTP settings of a user account.
+ */
+class TotpConfigForm extends Form
```
As a user I may get a new phone sooner or later. To transfer 2FA from one phone to another, I could imagine disabling it temporarily. But this may be not wanted for security or policy reasons. Then I have to add a second TOTP temporarily, but neither the code nor the schema permit this.

At least I'd allow replacing a TOTP atomically.

jrauh01 replied:
I hope you saved the QR codes on a different device than the one with your authenticator app on it, but I'm sure you have. To stay on topic, the user already has the option to screenshot the QR code themselves. Maybe we could add a hint like you mentioned in https://github.com/Icinga/icingaweb2/pull/5430#discussion_r2459880083.

Al2Klimov commented on library/Icinga/Authentication/IcingaTotp.php:
```diff
@@ -0,0 +1,157 @@
+        return $this;
+    }
+
+    public static function generate(string $user): static
```
> We will use the defaults of 6 digit tokens, 30 seconds interval and sha-1 algorithm for secret generation.

The default? So it's not strictly required? Good! Because SHA1 is insecure af and both Authy and Google support more than that.

(Bitte gleich gescheit machen.)

Al2Klimov commented on application/forms/Account/TotpConfigForm.php:
```diff
@@ -0,0 +1,217 @@
+
+    public function createElements(array $formData): void
+    {
+        if (! IcingaTotp::hasDbSecret($this->getDb(), $this->user->getUsername())) {
```
Isn't this DB query superfluous as you load the TOTP below anyway?

jrauh01 commented on schema/mysql.schema.sql:
```diff
@@ -64,5 +64,12 @@ CREATE TABLE icingaweb_schema (
+  `username` varchar(254) COLLATE utf8mb4_unicode_ci NOT NULL,
+  `secret`   varchar(255) NOT NULL,
+  `ctime`    timestamp NULL DEFAULT NULL,
+  PRIMARY KEY (`username`)
```
`mtime` is not needed, because the entry is never modified.

Al2Klimov replied:
Isn't it? What about replacement https://github.com/Icinga/icingaweb2/pull/5430#discussion_r2459873573? Shouldn't it be NULLed on removal, so that I can always see when I changed 2FA the last time?

If no, why is it in the Model then?

jrauh01 committed:
Add license headers for new files

jrauh01 committed:
Remove `@inheritDoc`

jrauh01 committed:
Add function return types

jrauh01 committed:
Add space after '!'

jrauh01 committed:
Adjust comment in `Auth::setupUser()`

jrauh01 committed:
Add function default values

...for `User::setTwoFactorEnabled()` and
`User::setTwoFactorSuccessful()`.

jrauh01 committed:
Remove unused `TotpConfigForm::onRequest()`

jrauh01 committed:
Remove redundant object creation

jrauh01 committed:
Don't show the stored secret anymore

jrauh01 committed:
Add property docs for model

jrauh01 committed:
Use microseconds timestamp for `ctime`

jrauh01 committed:
Add `icingaweb_totp` to pgsql schema

jrauh01 committed:
Add schema upgrades

jrauh01 committed:
Rename label to disable 2FA

Change the label to disable the 2FA from 'Remove TOTP Secret' to
'Disable 2FA'.

jrauh01 committed:
Use more appropriate names

jrauh01 committed:
Remove `TwoFactorTotp::setTotp()`

jrauh01 committed:
Remove superfluous variables

jrauh01 committed:
Add missing docs for `TwoFactorTotp`

jrauh01 committed:
Correct array indentation

jrauh01 committed:
Rename schema upgrades to '2.13.0.sql'

jrauh01 committed:
Rewrite `TwoFactorConfigForm` to ipl form

jrauh01 committed:
If no 2fa db table exists 2fa is disabled

jrauh01 committed:
Rewrite authentication forms to ipl forms

Removed description for `rememberme` input, because it wasn't
displayed anywhere.

jrauh01 committed:
Add `TotpTokenValidator`

It validates two things:
1. The token must only contain numbers
2. The token must have a specific length (this is configurable via
   the constructor, the default is 6)

jrauh01 committed:
Use one combined `LoginForm`

The form displays either the login inputs or the inputs to verify
the totp token depending on whether `'2fa_must_challenge_token'`
is set `true` in the session.

jrauh01 committed:
Improve `TwoFactorConfigForm`

A new class `FakeFormElement` is introduced which integrates
`ValidHtml` into IPL Web `CompatForm`s. The following parts of the
`TwoFactorConfigForm` are now added to the Form via the new class:

- The QR code image
- The manual auth URL for TOTP

The manual auth URL is now wrapped by a copy-to-clipboard element,
so the user doesn’t have to select it manually. Additionally the
QR code image and the manual auth url have now descriptions.

jrauh01 committed:
Use `endroid/qr-code` to generate QR code

Requires `endroid/qr-code` library in
Icinga/icinga-php-thirdparty#53.

Because we use **GPL-2.0** license for Icinga Web:

- `chillerlan/php-qrcode: **Apache-2.0** (incompatible)
- `endroid/qr-code`: **MIT** (compatible)

jrauh01 committed:
Add `CsrfCounterMeasure` to `TwoFactorConfigForm`

jrauh01 committed:
Add 'Download QR Code' action link

So the user can download the QR code to backup if they lose access
to their device.

The download currently does only work in chrome and firefox, but
not in safari. The reason is, that the action link uses
`ipl\Web\Url` which adds '///' between the schema and the base
path. The safari browser can't handle this.

'data:image/png;base64,...' is converted to
'data:///image/png;base64...'.

Al2Klimov reviewed (changes requested):
Tbh, I as a reviewer don't wanna even think about whose (branch) fault the red GHA is.

Just (someone) get it green, please. Thanks.

(Notes for the future myself:

<details>
  <summary>Test system (I use NixOS btw)</summary>

```nix
{ pkgs, ... }: {
  imports = [
    ./hardware-configuration.nix


  ];

  boot.tmp.cleanOnBoot = true;
  networking.hostName = "aklimov-2fa";
  networking.domain = "";
  services.openssh.enable = true;
  users.users.root.openssh.authorizedKeys.keys = [''ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIBIroHYGSRaRNFxlK90SS0aHwWjEME30pK5J1N/V1w6a aklimov@ws-aklimov7777777.local'' ];
  system.stateVersion = "23.11";

networking.firewall.allowedTCPPorts = [ 80 ];

virtualisation.oci-containers = {
backend = "podman";
containers.postgresql = {
autoStart = true;
image = "postgres:9.1";
environment = {
POSTGRES_PASSWORD = "123456";
};
ports = [ "127.0.0.1:5432:5432" ];
volumes = [ "postgres:/var/lib/postgresql/data" ];
};
};

  services.icingaweb2 = {
    enable = true;
    #virtualHost = "iw2.aklimov.net-dump.de";
    generalConfig.global.config_resource = "iw2";
    modules.monitoring.enable = false;
    authentications = {
      pg = {
        backend = "db";
        resource = "iw2";
      };
    };
    resources.iw2 = {
        type = "db";
        db = "pgsql";
        host = "127.0.0.1";
        port = "5432";
        dbname = "postgres";
        username = "postgres";
        password = "123456";
        charset = "utf8";
      };
    roles = {
      adm = {
        users = "icingaadmin,Alexander A. Klimov";
        permissions = "*";
      };
    };
    modulePackages = {
    };
  };

nixpkgs.overlays = [
(_: prev: {
icingaweb2-ipl = prev.icingaweb2-ipl.overrideAttrs (old: {
src = prev.fetchFromGitHub {
owner = "Icinga";
repo = "icinga-php-library";
rev = "v0.18.1";
hash = "sha256-V2PkRtVzMnzBpNLUMsSl5NUSnrycgEHGXLGilbhcGfw=";
};
}); })
(_: prev: {
icingaweb2-thirdparty = prev.icingaweb2-thirdparty.overrideAttrs (old: {
src = prev.fetchFromGitHub {
owner = "Icinga";
repo = "icinga-php-thirdparty";
#rev = "274212c7e02223687eacdfbc78c48181878ac910"; # feature/2fa
#hash = "sha256-K+NRUDnQjGG0Lmi5AOtftfueZM+VE6qkOWXF9qLzi/Y=";
rev = "1983c72cd3bbc0f4f0df5f464db4985417b4e73d"; # aklimov/feature/2fa
hash = "sha256-3CIAfhSr9GGhytOPnSfnnt/H6xWJOlQzyrZpux0VZbI=";
};
}); })
    (_: prev: { icingaweb2 = prev.icingaweb2.overrideAttrs (old: {
      src = prev.fetchFromGitHub {
        owner = "Icinga";
        repo = "icingaweb2";
        #rev = "62f10871e9e270789def9a3776656ff15d094013"; # 2fa-new
        #hash = "sha256-iCoCCEGLO3vf9VdbDWmecBMaRuUAKJc7f9AeNKLQv0A=";
rev = "e0e2a22fbeb90ae721b5d301af235467aa22d3ef"; # aklimov/2fa-new
hash = "sha256-0SaOkROWkc20j7MxGWsx5q4aSQoRTDaQazd2SJjCPbU=";
#rev = "70f29827f921f24b2afc0556c4800dfa55a2d0c1"; # main
#hash = "sha256-fCd5kyJk7LXdKkZS7kppPO1wfeNTOD4esliqvSAtyV4=";
      };

      patches = [
        ./opcache_reset.patch

        # https://github.com/Icinga/icingaweb2/issues/5427
        ./migrations-db-no-pw.patch
      ];

      # https://github.com/NixOS/nixpkgs/pull/380065
      installPhase = old.installPhase + "\ncp -ra schema $out";
    }); })
  ];

}
```
</details>

Al2Klimov reviewed (commented):
Almost forgot! The commit history got out of hand. Consider making one or a few large commits with shared authorship where applicable:

https://docs.github.com/de/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

Al2Klimov commented on application/forms/Security/RoleForm.php:
```diff
@@ -573,6 +573,9 @@ public static function collectProvidedPrivileges()
             ],
+            'user/two-factor-authentication'=> [
+                'description' => t('Allow 2FA configuration in the account preferences')
+            ],
```
> Yes, it should be controllable if the user is allowed to enable 2FA.

But why?

Al2Klimov replied:
And for disabling? (Company policy, GitLab also has this, ...)

Al2Klimov reviewed (changes requested):
@lippserd Please write your requirements on 2FA + API requests.

* They just shall not work? (What a pity.)
* They just shall work with user:pass? (Security hole, our forms [already allow](https://git.icinga.com/icingaweb2/icingaweb2-module-partnerportal/-/issues/151#note_28370) API requests implicitly!)
* OTP header required?
* OTP is to be appended to the password? (`username:password;otp`)
* ...

jrauh01 committed:
Prevent bypassing 2fa by changing username case

When using postgres db as authentication backend it was possible
to bypass 2fa by using e.g. `ADMIN` instead of `admin` as username.

This is now prevented by:
1. Ensuring there is only one secret per user no matter the case
2. Lowering all usernames for filtering in the database

jrauh01 committed:
Show only secret instead of whole otpauth url

This is more user friendly, because most authenticator apps just
want the secret for manual creation.

jrauh01 committed:
Make `ctime` column not nullable

jrauh01 committed:
Add CLI commands for 2FA

The following commands are added:
- `icingacli twofactor list` lists all users that have 2FA enabled.
- `icingacli twofactor disable [<user>]` disables 2FA for a user.

jrauh01 committed:
Remove permission `user/two-factor-authentication`

The user should not need to ask for more security.

jrauh01 committed:
Add hint to store the secret for recovery

If the device running the authenticator app is lost, the secret can
be used to set up an authenticator app on another device. But for
that the secret has to be stored on another device when setting up
2FA.

jrauh01 committed:
Add docs for 2FA

jrauh01 committed:
wip

jrauh01 committed:
wip: hookable config form

jrauh01 committed:
wip: hookable auth

jrauh01 committed:
Remove redundant `$skip2fa` parameter from `Auth::isAuthenticated()`

The parameter was only ever passed as true by the RememberMe path in
`loginAction()`, intending to let a cookie-based re-authentication bypass the
2FA gate. It was unnecessary for the following reason:

`RememberMe::authenticate()` already calls `setTwoFactorSuccessful()` on the
user before `setAuthenticated()` is called, so the gate condition evaluates to
false on its own.

jrauh01 committed:
Redirect directly to the target URL after successful 2FA verification

Previously, a successful 2FA login redirected back to the login action via
`Url::fromRequest()`, which would then detect the authenticated session and
issue a second redirect to the actual destination. This was an unnecessary extra
round-trip.

The `SUBMIT_VERIFY_2FA` path now calls createRedirectUrl() directly, matching
the behaviour of the `SUBMIT_LOGIN` path for users without 2FA enrolled.

jrauh01 committed:
Remove `enroll()` and `unenroll()` from `TwoFactor` interface

Both methods are only ever called from within `onSuccessEnrollmentForm()` of the
implementing class — making them public interface methods exposes an internal
storage detail as API surface. Removing them from the interface makes the
contract reflect actual usage: implementations handle persistence internally.

jrauh01 committed:
Delegate 2FA verification form assembly to hook implementations

The form element name constants (`TOKEN_INPUT`, `SUBMIT_VERIFY_2FA`,
`SUBMIT_CANCEL_2FA`) are moved out of LoginForm onto TwoFactor so all call sites
share a single source of truth.

Replace the `LoginForm`-owned `assembleTwoFactorElements()` method and the
`getChallengeFormValidators()` interface stub with a proper
`assembleVerificationForm(CompatForm $form)` contract on `TwoFactor`. Each hook
implementation now builds its own verification UI. For the element names the
constants from `TwoFactor` have to be used, so the login form's success handler
can retrieve the values.

jrauh01 committed:
Remove stale commented-out 2FA code from Auth

Delete leftover commented-out lines that referenced the old
`TwoFactorTotp::hasDbSecret()` approach for setting the 2FA enabled flag.
That logic has since moved to the login and `RememberMe` paths via
`TwoFactorHook::loadEnrolled()`.

jrauh01 committed:
Add `getUsername()` helper to `TwoFactorHook`

Introduce a protected `getUsername(): ?string` method that resolves the acting
user's name from `$this->user` (set during the login/verification flow by
`loadEnrolled()`) with a fallback to the session's authenticated user
(enrollment flow). Centralises a two-sided expression that was previously
duplicated across every implementation method that needed a username.

jrauh01 committed:
Set 2FA enabled flag during `RememberMe` authentication

When a user is re-authenticated via a remember-me cookie the user object was
never told whether 2FA is required, so the gate in `Auth::isAuthenticated()`
would always pass without a challenge. Call `TwoFactorHook::loadEnrolled()`
during the `RememberMe` login path — mirroring what the password login path
already does — so the flag is set correctly before the user is marked as
authenticated.

jrauh01 committed:
Move enroll/unenroll responsibility to `TwoFactorEnrollmentForm`

Replace `assembleEnrollmentForm()` and `onSuccessEnrollmentForm()` on the
`TwoFactor` interface with three narrower contracts:

- `assembleEnrollmentFormElements()` — add only the method-specific credential
  fields. Submit buttons are no longer the implementation's concern.
- `enroll(CompatForm $form): bool` — read the submitted credential, verify it,
  and persist it. Return false on verification failure.
- `unenroll(): void` — remove the stored credential.

`TwoFactorEnrollmentForm` now owns the enroll and unenroll submit buttons, calls
the appropriate method, and handles notifications and redirects centrally. This
keeps all implementations free of button naming, redirect logic, and user-facing
messages.

jrauh01 committed:
Scope enrollment elements in a per-method `FieldsetElement`

Wrap each hook's enrollment elements in a `FieldsetElement` named after the
method (via `getName()`). This scopes their POST keys to the method name, so
switching between methods in the type selector no longer cross-populates
elements that share the same name across implementations. This also ensures that
the implementation has only access to the elements added by itself.

Update `assembleEnrollmentFormElements()` and `enroll()` on the `TwoFactor`
interface to accept a `FieldsetElement` instead of a `CompatForm`. Also guard
the no-method-selected case with an early return so the enroll button is only
rendered once a method is chosen.

jrauh01 committed:
Link directly to `TwoFactorEnrollmentForm` in interface docs

jrauh01 committed:
fixup! Delegate 2FA verification form assembly to hook implementations

jrauh01 committed:
Move 2FA challenge form assembly back to `LoginForm`

Replace `assembleVerificationForm()` on the `TwoFactor` interface with
a generic `assembleTwoFactorElements()` method on `LoginForm` itself.
The token input, verify button, and cancel button are now owned by the
login form rather than each hook implementation.

Remove the `TOKEN_INPUT`, `SUBMIT_VERIFY_2FA`, and `SUBMIT_CANCEL_2FA`
constants from the `TwoFactor` interface and define them directly on
`LoginForm`. Update `AuthenticationController` to reference the new
location.

jrauh01 committed:
fixup! wip: hookable config form

jrauh01 committed:
Use variable to not redundantly generate server request

jrauh01 committed:
fixup! Rewrite authentication forms to ipl forms

jrauh01 committed:
Fix typo in login-orbs CSS selector

corresponding CSS to use the correct spelling `#orb-notifications`.

jrauh01 committed:
Add `LoginPage` widget to centralise login page and modernize to ipl

Extract the login page structure (logo, footer, social links, decorative orbs)
from `login.phtml` into a reusable ipl-html widget. Extending `HtmlDocument`
rather than `BaseHtmlElement` lets it emit `#login` and the seven `.orb` sibling
divs without a wrapping root tag.

Both the login action and the upcoming 2FA challenge action can now share the
same visual scaffolding without duplicating markup and without using view
scripts anymore.

jrauh01 committed:
Add `TwoFactorChallengeForm` for the 2FA verification step

Split the two-factor challenge out of `LoginForm` into its own standalone form
which handles token verification, remember-me cookie persistence, and the cancel
flow that purges the temporary session and returns to login.

jrauh01 committed:
Add `AuthenticationController::twofactorAction()`

Add a dedicated action for the 2FA verification step. The action guards against
direct access without a pending authentication, renders `TwoFactorChallengeForm`
inside the same `LoginPage` scaffold, and redirects to the originally requested
URL after successful verification.

`loginAction()` gains a matching guard: if a 2FA session is already in progress
(e.g. the user navigates back), it redirects straight to the challenge instead
of showing the login form again.

jrauh01 committed:
Migrate `AuthenticationController` to `CompatController`

Convert `AuthenticationController` from the legacy Zend Controller to
`CompatController`, dropping `login.phtml` in favour of the new `LoginPage`
widget and `addContent()`. The view-variable assignments are replaced by
`setTitle()` and `addContent(new LoginPage(...))`.

In `CompatController`, `$this->controls` is the tab bar area rendered above the
page content. When no tabs are added it still emits an empty
`<div class="controls">` wrapper. Replacing it with a bare `HtmlDocument` which
renders as nothing when empty suppresses that wrapper entirely, keeping the
login page markup clean.

Two structural fixes required by the changed DOM nesting:
- `login.less`: height `100%` -> `100vh` (`#login` is now inside `.content`
  which has no explicit height, so percentage inheritance breaks)
- `history.js`: `#layout > #login` -> `#layout #login` (direct-child selector
  breaks because `#login` is now a grandchild of `#layout` through `.content`)

jrauh01 committed:
Strip 2FA logic out of `LoginForm`

`LoginForm` is now responsible only for the first authentication factor. After a
successful credential check, users with 2FA enrolled are redirected to
`authentication/twofactor` where `TwoFactorChallengeForm` takes over.

Remove `assembleTwoFactorElements()`, the `SUBMIT_VERIFY_2FA` /
`SUBMIT_CANCEL_2FA` / `TOKEN_INPUT` constants, the `getResponse()` helper, and
the `2fa_must_challenge` session branch from `assemble()`. The `onSuccess()`
switch-on-button is replaced with a simple login-only handler.

Also removes the `ON_SENT` handler from `AuthenticationController` that handled
the cancel button. It referenced `SUBMIT_CANCEL_2FA` which is removed here, and
the responsibility moves to `TwoFactorChallengeForm` in the next commit.

jrauh01 committed:
Add `TwoFactorState` to centralise 2FA session handling

Introduce `TwoFactorState` as a dedicated wrapper around the session namespace
used during the two-factor authentication flow, replacing the scattered raw
session key accesses in `LoginForm`, `TwoFactorChallengeForm`, and
`AuthenticationController`.

`challenge()` opens the flow by storing the temporary user,
`completeChallenge()` closes it by cleaning up both the user and the pending
remember-me cookie, and `isChallenged()` provides a single readable guard for
the controller actions.

jrauh01 committed:
fixup! Add `LoginPage` widget to centralise login page and modernize to ipl

jrauh01 committed:
Separate `$form` and `$loginButtons` in `LoginPage`

Replace the flat `ValidHtml|array $formContent` parameter with typed
`CompatForm $form` and `LoginButtonForm[] $loginButtons` parameters. This gives
the widget semantic knowledge of what it renders and removes the
`array_filter(array_merge(...))` workaround from the `AuthenticationController`.

jrauh01 committed:
Log 2FA challenge and verification outcome

Emit an info line when a user is challenged and when they pass verification, and
a warning when the submitted token is rejected. All three log lines include the
method name so the 2FA backend in use is always visible in the logs.

jrauh01 committed:
Reject API requests from users with 2FA enrolled

Basic Auth cannot complete a 2FA challenge, so users with 2FA enrolled would
silently bypass it via the API. Detect enrollment in `authHttp()` and respond
with **403** and a JSON error message.

jrauh01 committed:
fixup! Strip 2FA logic out of `LoginForm`

jrauh01 committed:
Reuse `LoginForm::REDIRECT_URL` in `TwoFactorChallengeForm`

Remove the duplicate constant and reference the one defined in LoginForm.

jrauh01 committed:
Normalise naming conventions across 2FA form constants and element names

Rename constants and HTML element names to a consistent `twofactor_*/submit_*`
scheme, drop the now unused `SUBMIT_LOGIN` constant from `LoginForm`, and update
all references in the controllers.

jrauh01 committed:
Reformat `LoginForm::assemble()` to inline `addElement()` call style

jrauh01 committed:
fixup! Add `AuthenticationController::twofactorAction()`

jrauh01 committed:
Extract `withRedirect()` helper in `AuthenticationController`

Replace the repeated pattern of building a URL and conditionally forwarding the
redirect param with a single protected method.

jrauh01 committed:
wip: cancel button

jrauh01 committed:
Replace cancel button with a link-styled back-to-login button

Replaces the cancel submit button on the 2FA challenge form with a minimal
link-styled button showing an arrow-left icon and "Back to login" label. The new
`.btn-back-to-login-link` style strips all button chrome (background, padding,
fixed height) and only underlines the label text on hover, making it visually
unobtrusive next to the primary verify button.

jrauh01 committed:
Remove TOTP enroll styles from core — moved to icinga-totp module

jrauh01 committed:
Remove `FakeFormElement` from core — moved to icinga-totp module
